src/frontend/src/optimizer/plan_node/stream_project_set.rs

```diff
@@ -45,26 +45,39 @@ impl StreamProjectSet {
             .i2o_col_mapping()
             .rewrite_provided_distribution(input.distribution());
 
+        let mut out_watermark_columns = FixedBitSet::with_capacity(core.output_len());
+        let mut out_monotonicity_map = MonotonicityMap::new();
         let mut watermark_derivations = vec![];
         let mut nondecreasing_exprs = vec![];
-        let mut watermark_columns = FixedBitSet::with_capacity(core.output_len());
         for (expr_idx, expr) in core.select_list.iter().enumerate() {
+            let out_expr_idx = expr_idx + 1;
+
             use monotonicity_variants::*;
             match analyze_monotonicity(expr) {
+                Inherent(monotonicity) => {
+                    if matches!(monotonicity, NonDecreasing | Constant) {
+                        // We can only propagate non-decreasing/constant monotonicity, because we will produce
+                        // NULLs after all values of the column are consumed. Only non-decreasing/constant
+                        // monotonicity can hold even after appending several NULLs.
+                        out_monotonicity_map.insert(out_expr_idx, monotonicity);
+                    }
```

Comment:
Your are so meticulous 👍

Reply:
After "several NULLs" there will be non-NULL values again, where the monotonicity does not seems to hold any more. 🤔

Reply:
> After "several NULLs" there will be non-NULL values again, where the monotonicity does not seems to hold any more. 🤔

Yes, @xiangjinwu also mentioned that in offline discussion. I haven't got time to update this PR.

Reply:
Indeed. In that case, I suggest that we ignore the NULLs caused by expression errors here. This way, we would still derive monotonicity for scalar functions, but we would not be able to derive monotonicity for SRFs in ProjectSet.

Reply:
For now let's not derive any monotonicity for ProjectSet, leaving it for future.